app/src/main/java/com/quark/client/database/Users.kt

```diff
@@ -73,6 +77,51 @@ class Users(
                 continuation.resumeWithException(exception)
             }
     }
+
+    /**
+     * Takes a prompted username, then searches firebase to see if it exists, Used as a dependency
+     * for setUsername
+     * @param username: String
+     * @return Boolean
+     * @see UserProfile
+     */
+
+    private fun usernameUsedQuery(username: String): Boolean {
+        var result = true
+        firestore.collection("userprofiles").whereEqualTo("username", username).get()
+            .addOnSuccessListener { document ->
+                if (document == null) {
+                    result = false
+                }
+                else {
+                     result = true
+                }
+            }
+        return result
+    }
+
+
+    /**
+     * Takes in a username, and will set it to the current profile if there is a profile
+     * logged in, and if the username is not taken. Returns true on success, and false on failure
+     * @param username: String
+     * @param user: FirebaseUser
+     * @return Boolean
+     * @see UserProfile
+     */
+    suspend fun setUsername(username: String, user: FirebaseUser): Boolean? = suspendCancellableCoroutine {  continuation->
+        if(!usernameUsedQuery(username)) {
```

Comment:
do this in SignUp.kt before passing to setUsername instead, so that you can use LaunchedEffect